### PR TITLE
Bluetooth presence sensor

### PR DESCRIPTION
**Pull request by atrovato**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- ~~[ ] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)~~
- ~~[ ] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).~~
- ~~[ ] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).~~

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Allow to create Bluetooth device as presence sensor.

(temporary disable other theorical Bluetooth features).

Based on [presence-sensor PR](https://github.com/GladysAssistant/Gladys/pull/1024)

### CODE REVIEW

**Comment by Pierre-Gilles** [comment]
Just tried to test it, I got "reason: 'String PRESENCE-SENSOR does not represent a valid UUID'" and it made Gladys crash:

Screenshot 2020-12-21 at 21 27 28

**Comment by atrovato** [comment]
Wow, thanks... I'll look for this.

**Comment by atrovato** [comment]
Ok, this was due to notification auto-subscription on startup.
But as Bluetooth is not really functionnal for now, I remove this overflow function.
I'm already working on another version on Bluetooth devices, and still waiting for noble lib upgrades!

This sould be fixed now.

**Comment by Pierre-Gilles** [comment]
Thanks! I'm checking the new code :)

**Comment by Pierre-Gilles** [comment]
When I add a device in Gladys, then I remove the device, if I go back to the "setup" tab (I refresh the page in between), the device stays as "Update in Gladys", and not "Create in Gladys". why that? 

Screenshot 2020-12-22 at 20 35 43

**Comment by Pierre-Gilles** [comment]
I have a few questions on how it works :D 

When I did the integration on Gladys 3, the algorithm was:

```
Every XXX seconds:
    -> scan for bluetooth device during YY seconds
    -> foreach device found:
              -> device with address ZZ seen ! 
              -> IF (User is associated with address ZZ)
                     THEN -> User is seen ! 
```

In your case, what's the algorithm? I don't really grasp how you handle having different polling values per bluetooth device, and how you avoid having hundreds of scans at the same time in case the user registers 100 presence devices.

**Comment by atrovato** [comment]
> When I add a device in Gladys, then I remove the device, if I go back to the "setup" tab (I refresh the page in between), the device stays as "Update in Gladys", and not "Create in Gladys". why that?

When Bluetooth device is detected, I store it into a discovered map. I store it according to current device state in Gladys.
I have to change it to merge with current Gladys device state only on "get devices" and make it dynamic.

> In your case, what's the algorithm? I don't really grasp how you handle having different polling values per bluetooth device, and how you avoid having hundreds of scans at the same time in case the user registers 100 presence devices.

Mine is now only based on device poll delay. I think we could have problems with 100 devices.
I will change it to scan every XXX seconds as in V3. I will add a configuration page to change to scan delay.

I will need to add a token to allow device connection (which means stop scanning during this phase).

**Comment by Pierre-Gilles** [comment]
Agree with that! When testing locally, even with a few devices it was already "too much". I don't think the poll is adapted in this case, a global scan should be perfect and way more performant :)

**Comment by atrovato** [comment]
Good news (?), looks ready for testing and reviewing.

**Review by Pierre-Gilles — COMMENTED** [review]
Hi ! Thanks for this PR :) 

I try to do a first review from a user perspective, without reading the code.

When I arrive on the "setup" screen (which is more "presence" screen at the end)

Screenshot 2021-02-05 at 15 14 19

I have little knowledge of how this is useful. How does it work?

It feels like the screen is much a patchtwork of all options that are required by the backend, but not really something designed for the user 😄 I think we can improve that! 

- What is the reload button for?
- What is the "scan for device presences" for ?
- How does it work?

**Comment by atrovato** [comment]
> which is more "presence" screen at the end
For now yes, but at the end, all setup will be in this tab.

I can rename it at this time, and change later, when more setup will be available.

> It feels like the screen is much a patchtwork of all options that are required by the backend, but not really something designed for the user 😄 I think we can improve that!

Not sure to well understand. 
I'll add some explaination texts (and remove useless buttons).

But I don't really know how to get a better UI/UX. I'll ask on the forum for help.

**Comment by atrovato** [comment]
And like this?

image

File: front/src/config/i18n/fr.json [hunk 1]
```diff
@@ -348,6 +349,18 @@
           "cancelLabel": "Annuler",
           "successLabel": "Effectué"
         }
+      },
+      "setup": {
+        "title": "Scanner de présence",
+        "noConfigLabel": "La configuration n'est pas chargée, merci de réessayer.",
+        "errorLabel": "Une erreur est survenue lors du chargement de la configuration, merci de consulter les logs du navigateur.",
+        "presenceScannerDescription": "La fonctionnalité scanner de présence analyse les périphériques Bluetooth disponibles à interval régulier. Chaque fois qu'un appareil est détecté, un nouvel état est enregitré pour mémoriser la dernière fois qu'il a été vu.",
```
Comment (Pierre-Gilles) [inline front/src/config/i18n/fr.json]: *enregitré

File: front/src/config/i18n/fr.json [hunk 2]
```diff
@@ -348,6 +349,18 @@
           "cancelLabel": "Annuler",
           "successLabel": "Effectué"
         }
+      },
+      "setup": {
+        "title": "Scanner de présence",
+        "noConfigLabel": "La configuration n'est pas chargée, merci de réessayer.",
+        "errorLabel": "Une erreur est survenue lors du chargement de la configuration, merci de consulter les logs du navigateur.",
```
Comment (Pierre-Gilles) [inline front/src/config/i18n/fr.json]: Do we really want to advertise looking at the log? I've never seen this in any product

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
C'est mieux, mais je ne comprend toujours pas comment ça fonctionne

En tant qu'utilisateur, je cherche un résultat, pas une fonctionnalité

Si je me mets dans la peau d'un utilisateur qui possède un porte clé NUT bluetooth et qui veut être mis en "à la maison" quand il rentre et "hors de la maison" quand il part, honnêtement là il est perdu. Comment ça marche ?

Il faut que la page réponde à cette question entièrement, il ne faut pas décrire la fonctionnalité, il faut expliquer étape par étape comment faire.

Autre problème: tout le monde va penser que la détection de bluetooth ça fonctionne avec les téléphones, ce qui ne marche pas. Il faut être clair sur ce point aussi.

**Comment by atrovato** [comment]
En vrai cette page est simplement une page pour configuration la fréquence d'analyse de présence, ou pour désactiver la fonctionnalité. Je peux expliquer que la détection de la présence est liée à une feature présence.

Du coup, sur cette page, étant une page de setup, je ne suis pas sûr de comprendre comment lier des résultats. Sinon ajouter un cadre pour lister les devices présences et la date de dernière vue (qui risque d'ajouter beaucoup de lourdeur et de redondance).

> Si je me mets dans la peau d'un utilisateur qui possède un porte clé NUT bluetooth et qui veut être mis en "à la maison" quand il rentre et "hors de la maison" quand il part, honnêtement là il est perdu. Comment ça marche ?

Je ne peux pas répondre à cette question, car la on ne parle que de détection d'appareils, il n'y a donc pas de lien avec un utilisateur, qui sera dans les scènes comme nous l'avions dit. Je ne peux donc pas expliquer comment ça marche clairement? car ce n'est pas encore en place, me trompe-je ?.
Sinon à part dire qu'on analyse les appareils Bluetooth visibles à intervalle régulier, pareil je ne suis pas sûr de savoir comment compléter les informations.

Pour la détection des mobiles (qui ne se fait pas), je peux ajouter l'information, mais sur toutes les pages du Bluetooth alors, car que ce soit la découverte des appareils ou la configuration du service, cette information est légitime.
En fait, je pense que l'information, si il y a lieu d'être, devrait uniquement être sur la partie découverte, plus que sur la page configuration.
Ceci-dit, je pense qu'il est possible de faire la détection des smartphones, avec quelques manipulations, mais ce sera l'étude d'un autre sujet.

Je pense que je vais avoir besoin d'aide pour répondre à toutes ces questions, et intégrer tout ça de façon propre dans l'UI / UX.

@VonOx @Terdious @cicoub13 @Forum ?

**Comment by Pierre-Gilles** [comment]
> Je ne peux pas répondre à cette question, car la on ne parle que de détection d'appareils, il n'y a donc pas de lien avec un utilisateur, qui sera dans les scènes comme nous l'avions dit. Je ne peux donc pas expliquer comment ça marche clairement? car ce n'est pas encore en place, me trompe-je ?.

Ah, je pensais que cette PR permettait de faire de la détection de personnes en bluetooth de bout en bout.

En fait j'ai review la PR de manière fonctionnelle: qu'est ce que je peux faire avec cette PR en tant qu'utilisateur ?

Du coup est-ce que tu veux mettre en place la gestion de la présence dans les scènes ? 

Je peux m'en charger, mais ça sera pas pour tout de suite, je suis sur le multi-utilisateur pour l'instant de mon côté

**Comment by atrovato** [comment]
Oui, en effet, la n'est que le Bluetooth, afin de ne pas mélanger tous les sujets, et alourdir les revues.
Je peux tenter de regarder pour la partie des scenes, mais je ne garantis pas de ma disponibilité.
J'ai également changer la description dans la bulle info en m'inspirant des aides de la communauté.

**Comment by Pierre-Gilles** [comment]
Oui je comprend, c'est vrai que c'est plus facile de review une PR simple. Comme tu veux, dis moi si jamais tu t'y mets, sinon je m'y mettrais après le multi-utilisateurs :)

**Comment by Pierre-Gilles** [comment]
@atrovato j'ai fini le multi-utilisateurs donc je suis passé sur ça :) 

WIP: 
Screenshot 2021-02-15 at 18 08 12

**Comment by Pierre-Gilles** [comment]
Bon, la partie "retour à la maison" est hyper facile, mais la partie "départ de la maison" est pas simple...

Je ne sais pas si chercher à faire ça dans les scènes c'est pas hyper complexe pour l'utilisateur.

T'en penses quoi ?

**Comment by Pierre-Gilles** [comment]
J'ai mis dans cette PR mes quelques changements: https://github.com/GladysAssistant/Gladys/pull/1077

Je sais pas ce qu'on peut faire pour la partie "départ de la maison". Je pense il faut en discuter avant une implémentation

**Comment by atrovato** [comment]
Ah ok je vois qu'on rencontre le même problème, dans un sens c'est facile, mais l'autre... et je me vois mal créer 2 scenes...
Je pensais à ajouter une valeur sur le device, relative a un temps, et du coup mettre un timeout côté back pour checker au bout de n minutes si la user.last_seen_date est inférieure à n.
Mais on risque de se retrouver avec un max de timer côté back.
Sinon envoyez un device state si le scan ne trouve pas le device.

**Comment by Pierre-Gilles** [comment]
@atrovato on ne peut pas le faire "par device" car potentiellement l'utilisateur veut qu'on vérifie sa présence en se basant sur plusieurs devices. Genre un NUT sur ses clés + son bracelet Fitbit (je faisais ça perso sur Gladys v3), si au moins un des deux est détecté c'est que l'utilisateur est à la maison, et si les deux sont absents, c'est que l'utilisateur est parti.

J'ai pensé à une autre idée, une action dans les scène "vérifier présence utilisateur" qui ressemblerait à ça:

Screenshot 2021-02-19 at 09 25 47


L'utilisateur peut se créer une scène qui tourne toutes les 10 minutes et qui vérifie que aucun périphériques sélectionné n'est présent dans les précédentes XX minutes. Si un des périphériques est présent, on ne fait rien, sinon on met l'utilisateur comme absent de la maison sélectionnée.

Tu en pense quoi ?

**Comment by Pierre-Gilles** [comment]
Bon, j'ai commencé : 

Screenshot 2021-02-19 at 13 34 25

**Comment by Pierre-Gilles** [comment]
Merged in https://github.com/GladysAssistant/Gladys/commit/15b6027fdcb8d806f2d2ec6ce77bb58354e36ab1 ! :)

**Pull request closed without merging**